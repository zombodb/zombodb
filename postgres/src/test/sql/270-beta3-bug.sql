CREATE TABLE tas_people
(
  pk_ppl bigserial NOT NULL,
  ppl_fname text,
  ppl_lname text,
  CONSTRAINT tas_people_pkey PRIMARY KEY (pk_ppl)
);


CREATE TABLE tas_people_text
(
  pk_ppl_text bigserial NOT NULL,
  fk_ppl bigint,
  txt_full_text fulltext_with_shingles,
  CONSTRAINT tas_people_text_pkey PRIMARY KEY (pk_ppl_text)
);

insert into tas_people(ppl_fname, ppl_lname) values ('john', 'doe'), ('jane', 'snow'), ('eric', 'bridge'), ('mary', 'poppins'), ('larry', 'hopkins'), ('sammy', 'watkins');
insert into tas_people_text(fk_ppl, txt_full_text) values (1, 'some text'), (1, 'beer wine cheese'), (2, 'beer'), (3, 'flowers'), (4, 'minivan'), (5, 'some text'), (1, 'cola'), (1, 'water'), (null, 'beer'), (null, 'beer');


--drop index public.es_tas_people_text;
CREATE INDEX es_tas_people_text ON public.tas_people_text USING zombodb (zdb('public.tas_people_text', ctid), zdb(tas_people_text.*)) WITH (url='http://127.0.0.1:9200/', shards=4, replicas=1);
--drop index public.es_tas_people;
CREATE INDEX es_tas_people ON public.tas_people USING zombodb (zdb('public.tas_people', ctid), zdb(tas_people.*)) WITH (url='http://127.0.0.1:9200/', options='pk_ppl = <tas_people_text.es_tas_people_text>fk_ppl', shards=4, replicas=1, always_resolve_joins=true);

--DROP VIEW public.tas_people_view;
CREATE OR REPLACE VIEW public.tas_people_view AS
  select pk_ppl
    ,ppl_fname
    ,ppl_lname
    ,txt_full_text
    ,zdb('tas_people', tas_people.ctid) as zdb
  from tas_people
    left join tas_people_text on pk_ppl = fk_ppl;


select * from tas_people;
select * from tas_people_text;
select * from tas_people_view;

SELECT * from zdb_estimate_count('public.tas_people_text', 'txt_full_text:beer*');
SELECT * from zdb_estimate_count('public.tas_people_view', 'txt_full_text:beer*');

SELECT zdb_get_index_mapping('tas_people')->'mappings'->'data'->'_meta'->>'always_resolve_joins' = 'true';
ALTER INDEX es_tas_people SET (always_resolve_joins=false);
SELECT zdb_update_mapping('tas_people');
SELECT zdb_get_index_mapping('tas_people')->'mappings'->'data'->'_meta'->>'always_resolve_joins' = 'false';

DROP TABLE tas_people CASCADE;
DROP TABLE tas_people_text CASCADE;
