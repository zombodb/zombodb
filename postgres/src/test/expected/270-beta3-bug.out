CREATE TABLE tas_people
(
  pk_ppl bigserial NOT NULL,
  ppl_fname text,
  ppl_lname text,
  CONSTRAINT tas_people_pkey PRIMARY KEY (pk_ppl)
);
CREATE TABLE tas_people_text
(
  pk_ppl_text bigserial NOT NULL,
  fk_ppl bigint,
  txt_full_text fulltext_with_shingles,
  CONSTRAINT tas_people_text_pkey PRIMARY KEY (pk_ppl_text)
);
insert into tas_people(ppl_fname, ppl_lname) values ('john', 'doe'), ('jane', 'snow'), ('eric', 'bridge'), ('mary', 'poppins'), ('larry', 'hopkins'), ('sammy', 'watkins');
insert into tas_people_text(fk_ppl, txt_full_text) values (1, 'some text'), (1, 'beer wine cheese'), (2, 'beer'), (3, 'flowers'), (4, 'minivan'), (5, 'some text'), (1, 'cola'), (1, 'water'), (null, 'beer'), (null, 'beer');
--drop index public.es_tas_people_text;
CREATE INDEX es_tas_people_text ON public.tas_people_text USING zombodb (zdb('public.tas_people_text', ctid), zdb(tas_people_text.*)) WITH (url='http://127.0.0.1:9200/', shards=4, replicas=1);
--drop index public.es_tas_people;
CREATE INDEX es_tas_people ON public.tas_people USING zombodb (zdb('public.tas_people', ctid), zdb(tas_people.*)) WITH (url='http://127.0.0.1:9200/', options='pk_ppl = <tas_people_text.es_tas_people_text>fk_ppl', shards=4, replicas=1, always_resolve_joins=true);
--DROP VIEW public.tas_people_view;
CREATE OR REPLACE VIEW public.tas_people_view AS
  select pk_ppl
    ,ppl_fname
    ,ppl_lname
    ,txt_full_text
    ,zdb('tas_people', tas_people.ctid) as zdb
  from tas_people
    left join tas_people_text on pk_ppl = fk_ppl;
select * from tas_people;
 pk_ppl | ppl_fname | ppl_lname 
--------+-----------+-----------
      1 | john      | doe
      2 | jane      | snow
      3 | eric      | bridge
      4 | mary      | poppins
      5 | larry     | hopkins
      6 | sammy     | watkins
(6 rows)

select * from tas_people_text;
 pk_ppl_text | fk_ppl |  txt_full_text   
-------------+--------+------------------
           1 |      1 | some text
           2 |      1 | beer wine cheese
           3 |      2 | beer
           4 |      3 | flowers
           5 |      4 | minivan
           6 |      5 | some text
           7 |      1 | cola
           8 |      1 | water
           9 |        | beer
          10 |        | beer
(10 rows)

select * from tas_people_view;
 pk_ppl | ppl_fname | ppl_lname |  txt_full_text   |  zdb  
--------+-----------+-----------+------------------+-------
      1 | john      | doe       | some text        | (0,1)
      1 | john      | doe       | beer wine cheese | (0,1)
      2 | jane      | snow      | beer             | (0,2)
      3 | eric      | bridge    | flowers          | (0,3)
      4 | mary      | poppins   | minivan          | (0,4)
      5 | larry     | hopkins   | some text        | (0,5)
      1 | john      | doe       | cola             | (0,1)
      1 | john      | doe       | water            | (0,1)
      6 | sammy     | watkins   |                  | (0,6)
(9 rows)

SELECT * from zdb_estimate_count('public.tas_people_text', 'txt_full_text:beer*');
 zdb_estimate_count 
--------------------
                  4
(1 row)

SELECT * from zdb_estimate_count('public.tas_people_view', 'txt_full_text:beer*');
 zdb_estimate_count 
--------------------
                  2
(1 row)

SELECT zdb_get_index_mapping('tas_people')->'mappings'->'data'->'_meta'->>'always_resolve_joins' = 'true';
 ?column? 
----------
 t
(1 row)

ALTER INDEX es_tas_people SET (always_resolve_joins=false);
SELECT zdb_update_mapping('tas_people');
 zdb_update_mapping 
--------------------
 
(1 row)

SELECT zdb_get_index_mapping('tas_people')->'mappings'->'data'->'_meta'->>'always_resolve_joins' = 'false';
 ?column? 
----------
 t
(1 row)

DROP TABLE tas_people CASCADE;
NOTICE:  drop cascades to view tas_people_view
DROP TABLE tas_people_text CASCADE;
